packages/fuels-programs/src/receipt_parser.rs

```diff
@@ -0,0 +1,318 @@
+        contract_id: Option<&Bech32ContractId>,
+        output_param: &ParamType,
+    ) -> Result<Token> {
+        let contract_id = contract_id
+            .map(Into::into)
+            // During a script execution, the script's contract id is the **null** contract id
+            .unwrap_or_else(ContractId::zeroed);
+
+        let data = self
+            .extract_raw_data(output_param, &contract_id)
+            .ok_or_else(|| Self::missing_receipts_error(output_param))?;
+
+        ABIDecoder::decode_single(output_param, &data)
+    }
+
+    fn missing_receipts_error(output_param: &ParamType) -> Error {
+        error!(
+            InvalidData,
+            "ReceiptDecoder: failed to find matching receipts entry for {output_param:?}"
+        )
+    }
+
+    fn extract_raw_data(
+        &mut self,
+        output_param: &ParamType,
+        contract_id: &ContractId,
+    ) -> Option<Vec<u8>> {
+        match output_param.get_return_location() {
+            ReturnLocation::ReturnData if output_param.is_vm_heap_type() => {
+                self.extract_return_data_heap(contract_id)
+            }
+            ReturnLocation::ReturnData => self.extract_return_data(contract_id),
+            ReturnLocation::Return => self.extract_return(contract_id),
+        }
+    }
+
+    fn extract_return_data(&mut self, contract_id: &ContractId) -> Option<Vec<u8>> {
+        for (index, receipt) in self.receipts.iter_mut().enumerate() {
+            if let Receipt::ReturnData { id, data, .. } = receipt {
+                if id == contract_id {
+                    let data = std::mem::take(data);
+                    self.receipts.remove(index);
+                    return Some(data);
+                }
+            }
+        }
+        None
+    }
+
+    fn extract_return(&mut self, contract_id: &ContractId) -> Option<Vec<u8>> {
+        for (index, receipt) in self.receipts.iter_mut().enumerate() {
+            if let Receipt::Return { id, val, .. } = receipt {
+                if *id == *contract_id {
+                    let data = val.to_be_bytes().to_vec();
+                    self.receipts.remove(index);
+                    return Some(data);
+                }
+            }
+        }
+        None
+    }
+
+    fn extract_return_data_heap(&mut self, contract_id: &ContractId) -> Option<Vec<u8>> {
+        // If the output of the function is a vector, then there are 2 consecutive ReturnData
+        // receipts. The first one is the one that returns the pointer to the vec struct in the
+        // VM memory, the second one contains the actual vector bytes (that the previous receipt
+        // points to).
+        // We ensure to take the right "first" ReturnData receipt by checking for the
+        // contract_id. There are no receipts in between the two ReturnData receipts because of
+        // the way the scripts are built (the calling script adds a RETD just after the CALL
+        // opcode, see `get_single_call_instructions`).
```

Comment:
Hoping this assumption never breaks, otherwise we're in a world of pain, haha.

Reply:
I'm the OP of this assumption. It will not break unless either:
1- Sway doesn't return a vector using `RET` anymore
2- We change the script inside `get_single_call_instructions`. 
There are talks about changing the way vectors and more generally heap types are treated by the compiler so n°1 might happen.
Otherwise, this assumption is actually pretty safe IMO (famous last words lol).

Reply:
> Do you think we should improve the assumption's rationale in-code?

Nah, not much we can do here, really. We just have to keep an eye out for changes in the compiler that could trigger the breakage of this assumption.